File: c++/leetcode/1570-Dot_Product_of_Two_Sparse_Vectors-M.cpp
class SparseVector {
public:
    
    SparseVector(vector<int> &nums) {
        int n = nums.size();
        for (int i = 0; i < n; ++i) {
            if (nums[i]) {
                indicies_.push_back(i);
                nums_.push_back(nums[i]);
            }
        }
    }
    
    // Return the dotProduct of two sparse vectors
    int dotProduct(SparseVector& vec) {
        int n1 = indicies_.size(), n2 = vec.indicies_.size();
        int i = 0, j = 0;
        int dot = 0;
        while (i < n1 && j < n2) {
            if (indicies_[i] == vec.indicies_[j]) {
                dot += nums_[i++] * vec.nums_[j++];
            } else if (indicies_[i] < vec.indicies_[j]) {
                ++i;
            } else {
                ++j;
            }
        }
        return dot;
    }

private:
    vector<int> indicies_;
    vector<int> nums_;
};

File: python/leetcode/1570-Dot_Product_of_Two_Sparse_Vectors-M.py
class SparseVector:
    def __init__(self, nums: List[int]):
        v = []
        for i, n in enumerate(nums):
            if n:
                v.append((i, n))
        self._v = v

    # Return the dotProduct of two sparse vectors
    def dotProduct(self, vec: 'SparseVector') -> int:
        res = 0
        v0, v1 = self._v, vec._v
        m, n = len(v0), len(v1)
        i, j = 0, 0
        while i < m and j < n:
            if v0[i][0] > v1[j][0]:
                j += 1
            elif v0[i][0] < v1[j][0]:
                i += 1
            else:
                res += v0[i][1] * v1[j][1]
                i += 1
                j += 1
        return res
